integrators/symplectic_runge_kutta_nyström_integrator_test.cpp

```diff
@@ -69,51 +167,306 @@ TEST_F(SymplecticRungeKuttaNyströmIntegratorTest,
-
-  // Integrate forward with a BAB integrator.
-  FixedStepSizeIntegrator<ODE> const& bab_integrator =
-      BlanesMoan2002SRKN11B<Length>();
-  evaluations = 0;
+  std::vector<ODE::SystemState> solution;
+  ODE harmonic_oscillator;
+  harmonic_oscillator.compute_acceleration =
+      std::bind(ComputeHarmonicOscillatorAcceleration,
+                _1, _2, _3, nullptr);
+  IntegrationProblem<ODE> problem;
+  problem.equation = harmonic_oscillator;
+  ODE::SystemState const initial_state = {{q_initial}, {v_initial}, t_initial};
   problem.initial_state = &initial_state;
   problem.t_final = t_final;
-  solution.clear();
-  bab_integrator.Solve(problem, period / 10);
-  EXPECT_THAT(AbsoluteError(x_initial, solution.back().positions[0].value),
-              AllOf(Ge(4E-14 * Metre), Le(5E-14 * Metre)));
-  EXPECT_THAT(AbsoluteError(v_initial, solution.back().velocities[0].value),
-              AllOf(Ge(3E-7 * Metre / Second), Le(4E-7 * Metre / Second)));
-  EXPECT_EQ(t_final, solution.back().time.value);
-  EXPECT_EQ(steps, solution.size());
-  EXPECT_EQ(11 * steps + 1, evaluations);
+  ODE::SystemState final_state;
+  problem.append_state = [&final_state](ODE::SystemState const& state) {
+    final_state = state;
+  };
+
+  for (int i = 0; i < step_sizes; ++i, step /= step_reduction) {
+    integrator.Solve(problem, step);
+    Time const t = final_state.time.value - t_initial;
+    Length const& q = final_state.positions[0].value;
+    Speed const& v = final_state.velocities[0].value;
+    double const log_q_error = std::log10(
+        AbsoluteError(q / q_initial, Cos(ω * t)));
+    double const log_p_error = std::log10(
+        AbsoluteError(v / v_amplitude, -Sin(ω * t)));
+    if (log_q_error <= -13 || log_p_error <= -13) {
+      // If we keep going the effects of finite precision will drown out
+      // convergence.
+      break;
+    }
+    log_step_sizes.push_back(std::log10(step / Second));
+    log_q_errors.push_back(log_q_error);
+    log_p_errors.push_back(log_p_error);
+  }
+  double const q_convergence_order = Slope(log_step_sizes, log_q_errors);
+  double const q_correlation =
+      PearsonProductMomentCorrelationCoefficient(log_step_sizes, log_q_errors);
+  LOG(INFO) << "Convergence order in q : " << q_convergence_order;
+  LOG(INFO) << "Correlation            : " << q_correlation;
+
+  EXPECT_THAT(RelativeError(integrator.order, q_convergence_order),
+              Lt(0.02));
+  EXPECT_THAT(q_correlation, AllOf(Gt(0.99), Lt(1.01)));
+  double const v_convergence_order = Slope(log_step_sizes, log_p_errors);
+  double const v_correlation =
+      PearsonProductMomentCorrelationCoefficient(log_step_sizes, log_p_errors);
+  LOG(INFO) << "Convergence order in p : " << v_convergence_order;
+  LOG(INFO) << "Correlation            : " << v_correlation;
+  // SPRKs with odd convergence order have a higher convergence order in p.
+  EXPECT_THAT(
+     RelativeError(((integrator.order + 1) / 2) * 2,
```

Comment:
I would write this:

integrator.order + (integrator.order % 2)

It would be clearer.